File: problems/chapter04/mshibatatt/002.py

def tribo(N, memo):
    if N == 0 or N == 1: 
        memo[N] = 0
        return 0
    elif N == 2:
        memo[N] = 1
        return 1
    
    if memo[N] != -1: 
        return memo[N]
    
    memo[N] = tribo(N-1, memo) + tribo(N-2, memo)
    return memo[N]

def main():
    N = int(input())
    memo = [-1] * (N+1)

    output = tribo(N, memo)

    print(output)
    


if __name__=='__main__':
    main()    

File: problems/chapter04/miiitomi/002.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<int64_t> memo;

int64_t tribo(int n) {
    if (n == 0 || n == 1) {
        return 0;
    } else if (n == 2) {
        return 1;
    }

    if (memo[n] != -1LL) return memo[n];

    return memo[n] = tribo(n-1) + tribo(n-2) + tribo(n-3);
}

int main() {
    int N;
    cin >> N;
    memo.assign(N+1, -1LL);

    cout << tribo(N) << endl;
}
